src/librustc_typeck/collect.rs

```diff
@@ -2256,6 +2267,175 @@ fn predicates_from_bound<'tcx>(
+            }
+        },
+        t if t.contains("arm") => {
+            match simd_width {
+                // 32-bit arm does not support vectors with 64-bit wide elements
+                8 | 16 if simd_elem_width < 8 => Some("neon"),
+                _ => None,
+            }
+        },
+        t if t.contains("aarch64") => {
+            match simd_width {
+                8 | 16 => Some("neon"),
+                _ => None,
+            }
+        },
+        t if t.contains("powerpc") => {
+            match simd_width {
+                // 64-bit wide elements are only available in VSX:
+                16 if simd_elem_width == 8 => Some("vsx"),
+                16 if simd_elem_width < 8 => Some("altivec"),
+                _ => None,
+            }
+        },
+        t if t.contains("mips") => {
+            match simd_width {
+                16 => Some("msa"),
+                _ => None,
+            }
+        },
+        _ => None,
+    }
+}
+
+/// Returns true if the target-feature allows using the SIMD type on C FFI:
+fn simd_ffi_feature_check(
+    target: &str, simd_width: usize, simd_elem_width: usize, feature: &'static str
+) -> bool {
+    match target {
+        t if t.contains("x86") => {
+            // FIXME: see simd_ffi_min_target_feature
+            match simd_width {
+                8 => feature.contains("mmx")
+                    || feature.contains("sse")
+                    || feature.contains("ssse")
+                    || feature.contains("avx"),
+                16 => feature.contains("sse")
+                    || feature.contains("ssse")
+                    || feature.contains("avx"),
+                32 => feature.contains("avx"),
+                64 => feature.contains("avx512"),
+                _ => false,
+            }
+
+        },
+        t if t.contains("arm") => {
+            match simd_width {
+                // 32-bit arm does not support vectors with 64-bit wide elements
+                8 | 16 if simd_elem_width < 8 => feature.contains("neon"),
+                _ => false,
+            }
+        },
+        t if t.contains("aarch64") => {
+            match simd_width {
+                8 | 16 => feature.contains("neon"),
+                _ => false,
+            }
+        },
+        t if t.contains("powerpc") => {
+            match simd_width {
+                // 64-bit wide elements are only available in VSX:
+                16 if simd_elem_width == 8 => feature.contains("vsx"),
+                16 if simd_elem_width < 8 => feature.contains("altivec"),
+                _ => false,
+            }
+        },
+        t if t.contains("mips") => {
+            match simd_width {
+                16 => feature.contains("msa"),
+                _ => false,
+            }
+        },
+        _ => false,
+    }
+}
+
+fn simd_ffi_check<'a, 'tcx, 'b: 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId, ast_ty: &hir::Ty, ty: Ty<'b>,
+) {
+    if !ty.is_simd() {
```

Comment:
Nice catch, and no, we don't gate this today:

```rust
use std::arch::x86_64::__m128;

#[allow(improper_ctypes)]
extern "C" {
    //fn e(x: __m128); // ERROR
    pub fn a(x: A);
    pub fn b(x: B);
}

#[repr(transparent)] pub struct A(__m128);
#[repr(C)] pub struct B(__m128);
```

both of these are allowed on stable Rust, and both should be rejected.

Reply:
We also should gate foreign function definitions:

```rust
use std::arch::x86_64::__m128;

pub extern "C" fn foo(x: __m128) -> __m128 { x }
```

I'll try to fix all of these issues as part of this PR. We can do a crater run afterwards, and see if we need to change any of these to warn instead of error.

Reply:
Sounds good to me!